src/clr-angular/data/datagrid/providers/column-order-model.service.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable } from '@angular/core';
+import { ColumnOrdersCoordinatorService } from './column-orders-coordinator.service';
+import { DatagridHideableColumnModel } from '../datagrid-hideable-column.model';
+import { DomAdapter } from '../../../utils/dom-adapter/dom-adapter';
+
+/**
+ * This is a model service that's responsible for:
+ * 1. Sharing the flex order data between datagrid-column and reorderabble-droppable
+ * 2. Updating the flex order data and tells ColumnOrderCoordinatorService when it should broadcast
+ * 3. Returning models at the previous and next flex orders
+ */
+
+@Injectable()
+export class ColumnOrderModelService {
+  constructor(private columnOrderCoordinatorService: ColumnOrdersCoordinatorService, private domAdapter: DomAdapter) {}
+
+  public flexOrder: number;
+
+  public headerEl: any;
+
+  public hideableColumnModel: DatagridHideableColumnModel;
+
+  get columnGroupId() {
+    return this.columnOrderCoordinatorService.columnGroupId;
+  }
+
+  get isAtFirst(): boolean {
+    return this.flexOrder === 0;
+  }
+
+  get isAtEnd(): boolean {
+    return this.flexOrder === this.columnOrderCoordinatorService.orderModels.length - 1;
+  }
+
+  get isHidden(): boolean {
+    return this.hideableColumnModel && this.hideableColumnModel.hidden;
+  }
+
+  public dropReceived(dropData: any) {
+    // updates column orders
+    // broadcasts updated order changes
+    console.log(dropData);
+  }
+
+  get nextVisibleColumnModel(): ColumnOrderModelService {
+    // collect all columns that appears after this column
+    // that means collect all models with greater flex order value
+    const nextVisibleColumnModels = this.columnOrderCoordinatorService.orderModels.filter(
+      model => model.flexOrder > this.flexOrder && !model.isHidden
+    );
+
+    // current column could be the one that visually appears at the end;
+    if (nextVisibleColumnModels.length === 0) {
+      return;
+    }
+
+    // to find the next immediate visible columns model
+    // find the model with the smallest flexorder value from nextVisibleColumnModels
+    return nextVisibleColumnModels.reduce((smallestFlexOrderModel, currentModel) => {
+      if (smallestFlexOrderModel.flexOrder > currentModel.flexOrder) {
+        return currentModel;
+      }
+      return smallestFlexOrderModel;
+    });
+  }
+
+  get previousVisibleColumnModel(): ColumnOrderModelService {
```

Comment:
Nitpick: that's a bit of code duplication with the `nextVisibleColumnModel` method. If you want a fun little exercise, try to refactor this going a bit more functional in order to have less code and no duplication. 😉

Reply:
ok will try out a few things :)

Reply:
Did a bit of refactoring here 😛 Please take a look.

Reply:
That's better, but you went waaaay too flexible. Remember, this helper method you wrote is private so it only really needs to work for your two use case.

I would have gone for something like this:
```typescript
private sortAndFindNext(desc = false): ColumnOrderModelService | null {
  const sortedVisibleColumnModels = this.columnOrderCoordinatorService.orderModels
    .filter(model => !model.isHidden)
    .sort(model => model.flexOrder);
  const index = sortedVisibleColumnModels.findIndex(this);
  if (index >= 0 && index < sortedVisibleColumnModels.length - 1) {
    return sortedVisibleColumnModels[index+1];
  } else {
    return null;
  }
}

get nextVisibleColumnModel(): ColumnOrderModelService {
  return this.sortAndFindNext(false);
}

get previousVisibleColumnModel(): ColumnOrderModelService {
  return this.sortAndFindNext(true);
}
```

Reply:
I see. Yes, using `.sort()` is smart! I have opted for the one below:
```ts
 private findAdjacentVisibleModel(prev = false): ColumnOrderModelService {
    const filteredVisibleColumnModels = this.columnOrderCoordinatorService.orderModels
      .filter(model => (!model.isHidden && prev ? model.flexOrder < this.flexOrder : model.flexOrder > this.flexOrder))
      .sort(model => model.flexOrder);

    return prev ? filteredVisibleColumnModels[filteredVisibleColumnModels.length - 1] : filteredVisibleColumnModels[0];
  }
```